### PR TITLE
Fix for toolchain installation folder

### PR DESCRIPTION
**Pull request by Ricardicus**
I was trying to install the rust compiler with Xtensa support as well as the toolchain
under a different folder than my main rust folder (at ~/.cargo) and ran into some issues and needed this fix for it to work. 
Maybe it would be interesting for you to have it upstream as well?
Thank you for this project!

### CODE REVIEW

**Review by SergioGasquez — APPROVED** [review]
Thanks for your contribution, your changes look good to me! Also, I would like to let you know that `rust-build` installation scripts are feature-frozen, so I would recommend you switching to [`espup`](https://github.com/esp-rs/espup), where you can [modify the installation folder](https://github.com/esp-rs/espup#xtensa-rust-destination-path) with `CARGO_HOME` and `RUSTUP_HOME` as in `rustup`

**Review by SergioGasquez — CHANGES_REQUESTED** [review]
Hi!

Sorry, just tested your changes again without declaring `CARGO_HOME` and it fails with: 
```console
esp@496739c50564:~/sandbox/rust-build$ ./install-rust-toolchain.sh 
Processing configuration:
--build-target          = esp32,esp32s2,esp32s3
--cargo-home            = /home/esp/.cargo
--clear-cache           = YES
--esp-idf-version       = 
--export-file           = export-esp.sh
--extra-crates          = ldproxy cargo-espflash
--installation-mode     = install
--llvm-version          = esp-15.0.0-20221201
--minified-esp-idf      = NO
--minified-llvm         = YES
--nightly-version       = nightly
--rustup-home           = /home/esp/.rustup
--toolchain-version     = 1.67.0.0
--toolchain-destination = /home/esp/.rustup/toolchains/esp
/home/esp/.cargo/bin/rustup
nightly-x86_64-unknown-linux-gnu
Processing toolchain for x86_64-unknown-linux-gnu - operation: install
./install-rust-toolchain.sh: line 222: ./rust-1.67.0.0-x86_64-unknown-linux-gnu/install.sh: No such file or directory
```

While using the script in the main branch works fine.

**Comment by Ricardicus** [comment]
> Hi!
> 
> Sorry, just tested your changes again without declaring `CARGO_HOME` and it fails with:
> 
> ```
> esp@496739c50564:~/sandbox/rust-build$ ./install-rust-toolchain.sh 
> Processing configuration:
> --build-target          = esp32,esp32s2,esp32s3
> --cargo-home            = /home/esp/.cargo
> --clear-cache           = YES
> --esp-idf-version       = 
> --export-file           = export-esp.sh
> --extra-crates          = ldproxy cargo-espflash
> --installation-mode     = install
> --llvm-version          = esp-15.0.0-20221201
> --minified-esp-idf      = NO
> --minified-llvm         = YES
> --nightly-version       = nightly
> --rustup-home           = /home/esp/.rustup
> --toolchain-version     = 1.67.0.0
> --toolchain-destination = /home/esp/.rustup/toolchains/esp
> /home/esp/.cargo/bin/rustup
> nightly-x86_64-unknown-linux-gnu
> Processing toolchain for x86_64-unknown-linux-gnu - operation: install
> ./install-rust-toolchain.sh: line 222: ./rust-1.67.0.0-x86_64-unknown-linux-gnu/install.sh: No such file or directory
> ```
> 
> While using the script in the main branch works fine.

ok!

Maybe it has to do with the environment... The line that fails there is because it cannot find
a file that it should have extracted from a tar archive. It seems to not have downloaded the
rust-1.67.0.0-x86_64-unknown-linux-gnu.tar.xz and extracted it in the local directory, or
somehow the archive just lies there, unextracted.

Maybe the download was interrupted somehow?

I am not sure if it has to do with my change (but that is what they all say).

It shouldn't matter if CARGO_HOME is defined or not, by default it should be set and exported to ~/.cargo. 
The CARGO_HOME variable is already set in the script, first locally on line 9:

`9: CARGO_HOME="${CARGO_HOME:-${HOME}/.cargo}"`

And if -c is used, the CARGO_HOME value is overwritten.

The rest was just to source the env-file under the cargo home directory if it was present.
I noticed that the overwriting mechanism didn't work there, if ~/.cargo/env was present,
it would be used even if specified otherwise with -c. 

I can install it both ways, so I am in the dark here..

**Comment by Ricardicus** [comment]
I reproduced the error you experienced, and it happens if I exit the download of the tar archives.
It was there before, but I think I have a fix for it now :)

File: install-rust-toolchain.sh [hunk 1]
```diff
@@ -158,17 +158,12 @@ function install_rust_toolchain() {
...
+    if [[ -e "${CARGO_HOME}/env" ]]; then
+        source "${CARGO_HOME}/env"
     else
-        if [[ -n "${CARGO_HOME}" ]] && [[ -e "${CARGO_HOME}/env" ]]; then
-            source ${CARGO_HOME}/env
-        else
-            echo "Warning: Unable to source .cargo/env"
-            export CARGO_HOME="${HOME}/.cargo"
-        fi
+        echo "Warning: Unable to source ${CARGO_HOME}/env"
```
Comment (SergioGasquez) [inline install-rust-toolchain.sh:164]:
```suggestion
        echo "Warning: Unable to source ${CARGO_HOME}/env. Verify that the path specified for ${CARGO_HOME} exists"
```

I would suggest adding a hint, the user can specify a wrong CARGO_HOME

**Comment by SergioGasquez** [comment]
> I reproduced the error you experienced, and it happens if I exit the download of the tar archives. It was there before, but I think I have a fix for it now :)

Thanks for taking care of it!

**Comment by Ricardicus** [comment]
thank you for letting me program my esp32 with Rust!

**Pull request merged**